src/clr-angular/forms/input/input-container.spec.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import {Component} from "@angular/core";
+import {TestBed} from "@angular/core/testing";
+import {FormsModule, NgControl} from "@angular/forms";
+import {By} from "@angular/platform-browser";
+
+import {ClrIconModule} from "../../icon/icon.module";
+import {ClrCommonFormsModule} from "../common/common.module";
+import {ControlService} from "../common/providers/control.service";
+
+import {ClrInput} from "./input";
+import {ClrInputContainer} from "./input-container";
+
+@Component({
+    template: `
+    <clr-input-container>
+        <input type="input" name="test" clrInput minlength="5" [(ngModel)]="model" />
+        <label>Hello World</label>
+        <clr-control-helper>Helper text</clr-control-helper>
+        <clr-control-error>Must be at least 5 characters</clr-control-error>
+    </clr-input-container>
+    `
+})
+class SimpleTest {
+}
+
+export default function(): void {
+    describe("ClrInputContainer component", () => {
+        let fixture, inputContainerDE, inputContainer, inputContainerEl;
+
+        beforeEach(() => {
+            TestBed.configureTestingModule({
+                imports: [ClrIconModule, ClrCommonFormsModule, FormsModule],
+                declarations: [ClrInputContainer, ClrInput, SimpleTest],
+                providers: [NgControl]
+            });
+            fixture = TestBed.createComponent(SimpleTest);
+            fixture.detectChanges();
+
+            inputContainerDE = fixture.debugElement.query(By.directive(ClrInputContainer));
+            inputContainer = inputContainerDE.componentInstance;
+            inputContainerEl = inputContainerDE.nativeElement;
+        });
+
+        it("injects the ControlService and subscribes in AfterContentInit", () => {
+            expect(inputContainerDE.injector.get(ControlService, null)).toBeTruthy();
+            expect(inputContainer.subscription).toBeTruthy();
+        });
+
+        it("projects the label", () => {
+            const label = inputContainerEl.querySelector("label");
+            expect(label).toBeTruthy();
+            expect(label.previousElementSibling).toBeFalsy();
+        });
+
+        it("projects the input", () => {
+            expect(inputContainerEl.querySelector("[clrInput]")).toBeTruthy();
+        });
+
+        it("projects the helper text", () => {
```

Comment:
I would make this test slightly more complicated to test `it("projects the helper text when the input isn't invalid")`. Double negation, I know, but "when then input is valid" doesn't describe it correctly.